File: spendy_backend/src/main/java/com/rohitcodes/expense_tracker/service/TransactionService.java
package com.rohitcodes.expense_tracker.service;

import com.rohitcodes.expense_tracker.dto.transaction.TransactionRequest;
import com.rohitcodes.expense_tracker.dto.transaction.TransactionResponse;
import com.rohitcodes.expense_tracker.entity.TransactionType;

import java.time.LocalDate;
import java.util.List;

public interface TransactionService {

    TransactionResponse createTransaction(Long userId, TransactionRequest request);

    List<TransactionResponse> getTransactions(
            Long userId,
            TransactionType type,
            LocalDate startDate,
            LocalDate endDate
    );
    List<TransactionResponse> getRecentTransactions(Long userId);

    TransactionResponse getTransactionById(Long userId, Long transactionId);

    TransactionResponse updateTransaction(Long userId, Long transactionId, TransactionRequest request);

    void deleteTransaction(Long userId, Long transactionId);
}

File: spendy-mobileFrontend/src/services/transactionService.ts
import { api } from "@/src/services/api";
import type {
  ImportStatus,
  Transaction,
  TransactionSource,
  TransactionType
} from "@/src/types/transaction";
export type TransactionPayload = {
  categoryId: number;
  amount: number;
  type: TransactionType;
  source: TransactionSource;
  importStatus: ImportStatus;
  transactionDate: string;
  description: string;
  merchant?: string;
  externalMessageId?: string;
  rawMessage?: string;
};
export const transactionService = {
  async getTransactions(
    userId: number,
    params?: {
      type?: TransactionType | "ALL";
      startDate?: string;
      endDate?: string;
    }
  ) {
    const response = await api.get<Transaction[]>(`/users/${userId}/transactions`, {
      params: {
        type: params?.type && params.type !== "ALL" ? params.type : undefined,
        startDate: params?.startDate,
        endDate: params?.endDate
      }
    });
    return response.data;
  },
  async getTransactionById(userId: number, transactionId: number) {
    const response = await api.get<Transaction>(`/users/${userId}/transactions/${transactionId}`);
    return response.data;
  },
  async createTransaction(userId: number, payload: TransactionPayload) {
    const response = await api.post<Transaction>(`/users/${userId}/transactions`, payload);
    return response.data;
  },
  async updateTransaction(userId: number, transactionId: number, payload: TransactionPayload) {
    const response = await api.put<Transaction>(
      `/users/${userId}/transactions/${transactionId}`,
      payload
    );
    return response.data;
  },
  async deleteTransaction(userId: number, transactionId: number) {
    await api.delete(`/users/${userId}/transactions/${transactionId}`);
    return transactionId;
  }
};